lib/dockerfile.ts

```diff
@@ -0,0 +1,226 @@
+						'Dockerfiles containing the ADD instruction are not supported. Please use COPY.',
+					);
+				case 'COPY':
+					// If there are any commands which have occured already, save them as an action
+					// group
+					if (commands.length > 0) {
+						this.actionGroups.push({
+							workDir,
+							commands,
+							fileDependencies: lastCopy,
+						});
+						commands = [];
+					}
+
+					lastCopy = Dockerfile.copyArgsToFileDeps(workDir, entry.args);
+					break;
+				case 'WORKDIR':
+					// Take every command we currently have, and save it as an action group,
+					// as encountering another working directory means that every command
+					// following must execute in a different action group
+					if (commands.length > 0) {
+						this.actionGroups.push({
+							workDir,
+							commands,
+							fileDependencies: lastCopy,
+						});
+					}
+					commands = [];
+					if (!_.isString(entry.args)) {
+						throw new DockerfileParseError('Non-string argument to WORKDIR');
+					}
+					workDir = entry.args;
+					break;
+				case 'RUN':
+					// Add this to the current set of commands
+					commands.push(Dockerfile.processRunArgs(entry.args));
+					break;
+			}
+		}
+
+		// Store any commands left over
+		if (commands.length > 0) {
+			this.actionGroups.push({
+				workDir,
+				commands,
+				fileDependencies: lastCopy,
+			});
+		}
+	}
+
+	// TODO: We can probably cache this value
+	/**
+	 * Get a list of action groups which should be re-executed if any of
+	 * the files passed to this function have changed.
+	 *
+	 * @param files A list of files whose contents have changed. The path
+	 * should be relative to the build context
+	 */
+	public getActionGroupsFromChangedFiles(
+		files: string[],
+	): DockerfileActionGroup[] {
+		let actionGroupIdx = 0;
+		for (const actionGroup of this.actionGroups) {
+			// FIXME: COPY src/ src/ won't work due to the fact that there isn't a glob on the end of
+			// the target - we can fix this by silenty converting src/ to src/*
+
+			// For every dependency, see if it matches the current action group
+			const matches = _(actionGroup.fileDependencies)
+				.flatMap(({ localPath }) => minimatch.match(files, localPath))
+				.uniq()
+				.value();
+
+			// If any of the files have changed we need to return all
+			// action groups which follow, as they could depend on either
+			// one of the trigger files, or the output of a previous action
+			// group
+			if (matches.length > 0) {
+				return this.actionGroups.slice(actionGroupIdx);
+			}
+
+			++actionGroupIdx;
+		}
+
+		return [];
+	}
+
+	private static copyArgsToFileDeps(
+		workDir: string,
+		copyArgs: string[],
+	): FileDependency[] {
+		if (!_.isArray(copyArgs)) {
+			throw new DockerfileParseError('Non-array arguments passed to COPY');
+		}
+
+		// Remove any flags
+		copyArgs = Dockerfile.removeDashedArgs(copyArgs);
+
+		if (copyArgs.length < 2) {
+			throw new DockerfileParseError('Incorrect argument count for COPY');
```

Comment:
It would be nice to use something like "Incorrect argument count for COPY, a minimum of 2 are required" to be a bit more helpful